### DISCUSSION TITLE
Field validation that depends on the values of sibling/parent/cousin etc. fields

### DISCUSSION THREAD

**Discussion by willd-mwp** [colinhacks/zod#988] [Q&A]
I've been evaluating `zod` as an alternative to `yup` which is the validation library we currently use.

The type for my form state contains a property `height`, such as this (other properties of `FormState` have been omitted for brevity):
```typescript
type FormState = { 
    height: {
        min: {
            isApplied: boolean,
            value: number
        },
        max: {
            isApplied: boolean,
            value: number
        },
    }
}
```

`height.min.value` has some validation rules, but some of these rules are dependent on other properties, but I'm not sure of what the recommended approach for implementing them with `zod`.
1. if `height.min.isApplied` is true, we want to apply a `number().min().max()`, other we just want to apply `number()`
    - In `yup`, I achieve this using the `when()` method, which allows referencing the value of a sibling field
        ```ts
        value: number().when("isApplied", {
          is: true,
          then: number().required().min(0).max(100),
          otherwise: number()
        })
        ```
   - In `zod`, there doesn't seem to be an equivalent of `when()`, and it's not clear how i could utilise any of the existing schema methods to achieve this behaviour. Maybe using `union()` somehow?

1. if `height.min.isApplied` is true **AND**  `height.max.isApplied` is true, we want to apply an additional constraint that checks that `height.min.value` is less than `height.max.value`
    - In `yup`, I would acheive this using the `test()` method, which allows you to construct a custom validation check via a callback that has access to parent properties via the `from` array property of the `ctx` parameter of the callback
        ```ts
        number().required().min(0).max(100).test("less-than-maximum", "Min height must be less than max height",
            (value, ctx: any) => {
                const height = ctx.from[2].value.height;
                return value !== undefined && value !== null && (!height.max.isApplied || value < height.max.value);
            })
        ```
   - In `zod`, there's a `refine()` and `superRefine()` method for doing custom validation checks, but the `ctx` parameter of the `superRefine()` callback doesn't seem to give you access to the object tree like `yup` does, so I don't know how you'd reference `height.max.isApplied` and `height.max.value` from the schema applied to `height.min.value`

Overall I'm really liking the library, especially it's leverage of TypeScript functionality such as with discriminated unions in `ZodIssue`. 
If anyone could suggest how `zod` can be used to solve the above problems, I don't think I'd have any problems recommending it over `yup`.

**Comment by JacobWeisenburger** [comment]
Would this work for you?

Let me know if you have any questions.

```ts
const heightSchema = z.object( {
    value: z.number(),
    min: z.number().optional(),
    max: z.number().optional(),
} ).superRefine( ( input, ctx ) => {
    const { value, min, max } = input

    if ( min != null && value < min ) {
        ctx.addIssue( {
            code: z.ZodIssueCode.too_small,
            minimum: min,
            type: 'number',
            inclusive: true,
            message: `Height must be greater than ${ min }`,
        } )
    }

    if ( max != null && value > max ) {
        ctx.addIssue( {
            code: z.ZodIssueCode.too_big,
            maximum: max,
            type: 'number',
            inclusive: true,
            message: `Height must be less than ${ max }`,
        } )
    }
} )

console.log( heightSchema.safeParse( { value: 1 } ).success ) // true
console.log( heightSchema.safeParse( { value: 1, min: 1, max: 10 } ).success ) // true

console.log( heightSchema.safeParse( { value: 1, min: 2 } ).success ) // false
console.log( heightSchema.safeParse( { value: 11, max: 10 } ).success ) // false
console.log( heightSchema.safeParse( { value: 1, min: 2, max: 10 } ).success ) // false
console.log( heightSchema.safeParse( { value: 11, min: 2, max: 10 } ).success ) // false
```

**Reply by willd-mwp** [reply]
@JacobWeisenburger 

The `{ isApplied: boolean, value: number}` structure is a type in our graphql schema, for which I've used `graphql-codegen` to generate typescript typings. The structure of that is kept for use in the form state to keep the components doing lots of mapping to and from the types that the graphql schema accepts.
This type structure creates this nesting of related values, so I need to ability to navigate up and down the validation tree in the `superRefine()` callback

**Reply by JacobWeisenburger** [reply]
What do you mean by "validation tree"? The first arg of the superRefine callback function (`input`) is whatever was passed to parse/safeParse. I think this is what you mean, but please clarify if I am not understanding.

**Reply by willd-mwp** [reply]
@JacobWeisenburger 

So if I have my `FormState` type:
```typescript
type FormState = { 
    height: {
        min: {
            isApplied: boolean,
            value: number
        },
        max: {
            isApplied: boolean,
            value: number
        },
    }
}
```

And a validation schema for that `FormState` type:
```typescript
const schema: z.ZodSchema<FormState> = z.object({
    height: z.object({
        min: z.object({
            isApplied: z.boolean(),
            value: z.number().required().min(0).max(100).superRefine((val, ctx) => {
                console.log("val: ", val);
                console.log("ctx: ", ctx);
            })
        }),
        max: z.object({
            isApplied: z.boolean(),
            value: z.number().required().min(0).max(100)
        })
    })
}) 
```

And then a `react-hook-form` form using that validation schema that's initialised with some default values
```typescript
const {
    register,
    formState: { touchedFields, errors, isDirty, isValid, isSubmitting },
    handleSubmit,
    reset,
    control,
  } = useForm<FormState>({
    mode: "onBlur",
    defaultValues: {
        height: {
            min: {
                isApplied: true;
                value: 20
            },
            max: {
                isApplied: false;
                value: 120
            },
        }
    },
    resolver: zodResolver(schema),
  });
```  

In my dev console I'm going to get this output from the console logging that I've put into the `superRefine()` function on `height.min.value`
```text
value:  20
ctx: {addIssue: ƒ}
```

So this demonstrates that in the `superRefine` function, I only have access to the value that I attached the `superRefine()` to, and not the whole `height` object.

Compare this to using a `test()` function in `yup`
```typescript
const schema: yup.ObjectSchema<FormState> = yup.object({
    height: yup.object({
        min: yup.object({
            isApplied: yup.boolean(),
            value: yup.number().required().min(0).max(100)
            			.test("example", "example error", (val, ctx) => {
			                  console.log("val: ", val);
			                  console.log("ctx: ", ctx);
		         	     })
        }),
        max: yup.object({
            isApplied: yup.boolean(),
            value: yup.number().required().min(0).max(100)
        })
    })
}) 
```

Where in the dev console, I get this output
```text
val:  20
ctx: {
	  from: [
		{schema: ObjectSchema, value: 20}
		{schema: ObjectSchema, value: {isApplied: true, value: 20}}
		{schema: ObjectSchema, value: {min:{isApplied: true, value: 20}, max:{isApplied: false, value: 120}}
		{schema: ObjectSchema, value: {height:{min:{isApplied: true, value: 20}, max:{isApplied: false, value: 120}}}
	  ]
	  parent: {isApplied: true, value: 20}
	  path: "height.min.value"
	  schema: NumberSchema {type: 'number', _whitelist: ReferenceSet, _blacklist: ReferenceSet, internalTests: {…}, _typeCheck: ƒ, …}
	  type: "example"
	  ... (there's some other properties but I omitted them for brevity)
}	   
```

Where you can see, that via the array in the `from` property, I can access any property in `FormState`, which allows me to reference `height.max.isApplied` and `height.max.value` from a `test()` function applied to `height.min.value`.
This isn't possible in zod's `superRefine()`, unless I were to declare the `superRefine()` on `height` instead of `height.min.value`, but then `zodResolver` wouldn't be putting any `ZodErrors` resulting from the `superRefine()` at `errors.height.min.value` where I want it.

**Reply by willd-mwp** [reply]
@JacobWeisenburger 
Has that made my use case clearer?

**Reply by lveillard** [reply]
@willd-i2o did you manage to fix this?

**Reply by willd-mwp** [reply]
@lveillard 

no. because I couldn't find a good way to implement a solution for this scenario in `zod` , I gave up up on my spike of `zod` and continued using `yup`

**Reply by datner** [reply]
Sorry to hear that, but you only had to invoke the `superRefine` on the schema that is part of the interaction instead of the schema of the number

```ts
const schema: z.ZodSchema<FormState> = z.object({
    height: z.object({
        min: z.object({
            isApplied: z.boolean(),
            value: z.number().required().min(0).max(100)
        }),
        max: z.object({
            isApplied: z.boolean(),
            value: z.number().required().min(0).max(100)
        })
    })
    .superRefine((val, ctx) => {
        console.log("val: ", val);
        console.log("ctx: ", ctx);
    })
}) 
```

now val will contain 
```js
{
  min: {
    isApplied: boolean,
    value: number
  },
  max: {
    isApplied: boolean,
    value: number
  }
}
```

---------------------

its a different way of thinking, where instead of every schema being an entire cohesive thing, it's made up of many tiny schemas, allowing you to reuse most of you schemas and modify them per-case if they don't fit exactly

```ts
const Dimension  = z.object({
  isApplied: z.boolean(),
  value: z.number()
})
.refine(({ isApplied, value}) => !isApplied || value <= 100 && value >= 0, "example error")

// or using union like you suggested

const Dimension = z.discriminatedUnion('isApplied', [
  z.object({ isApplied: z.literal(false), value: z.number().default(0) }),
  z.object({ isApplied: z.literal(true). value: z.number().min(0),max(100) })
])

const Form: z.ZodSchema<FormState> = z.object({
  height: z.object({
    min: Dimension,
    max: Dimension
  })
  .superRefine(({min, max}, ctx) => {
    if (min > max) {
      ctx.addIssue({
            code: z.ZodIssueCode.too_small,
            minimum: min,
            type: 'number',
            inclusive: true,
            path: ['max', 'value'],
            message: `Max value can't be lower than min value`,
      })
    }
})
```

**Reply by datner** [reply]
@willd-i2o

**Reply by datner** [reply]
@lveillard

**Reply by willd-mwp** [reply]
@datner 

That does means that when something like `zodResolver` for `react-hook-form` applies the validation, the error would appear on `height` rather than on `height.min.value` where I would want it

**Reply by datner** [reply]
@willd-i2o 
only if you don't supply the `path` variable 😄 
```ts
ctx.addIssue({
            code: z.ZodIssueCode.too_small,
            minimum: min,
            type: 'number',
            inclusive: true,
            // The following property controls the "location" of the error
            path: ['max', 'value'], // <-- error will register under height.max.value
            message: `Max value can't be lower than min value`,
      })
```

**Reply by willd-mwp** [reply]
@datner 

Ah, that's useful, thanks.
